streams/src/main/java/org/apache/kafka/streams/processor/To.java

```diff
@@ -65,4 +67,22 @@ public To withTimestamp(final long timestamp) {
         this.timestamp = timestamp;
         return this;
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final To to = (To) o;
+        return timestamp == to.timestamp &&
+            Objects.equals(childName, to.childName);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(childName, timestamp);
```

Comment:
Needed to add this back, because we use it for testing now... Thoughts?

Reply:
We use `To` in hash-collections in our tests? It seems like we could work around that in other ways.

Reply:
We don't use it explicitly, but it's required for mocking. Eg. https://github.com/apache/kafka/pull/6645/files#diff-2eb683696aa96820098ed11941833ee3R36

Reply:
Thanks for the context. I think that mock actually only depends on Equals. At least, I replaced this with `throw new UnsupportedOperationException();`, and the test still passes for me.

Reply:
Note that it would be equally safe (and maybe a little more intuitive) to just make the fields `final`. This results in a bigger change, though, because it needs changes in ProcessorContextImpl. That might not be a bad thing.

It seems like the `To` class is mutable primarily so that ProcessorContextImpl can maintain an _immutable_ reference to it. But there's really no increase in safety between an immutable reference to a mutable object vs a mutable reference to an immutable object. Arguably, the latter is a little better because `To` is a data container (so immutability is best), whereas `ProcessorContextImpl` is a full-blown behavioral object that already has a bunch of mutable references to things.

Reply:
Ack about throwing an exception.

I am open to the other refactoring, but I won't do it in this PR. Feel free to do a MINOR PR directly or create a ticket. I just don't want to convolute this PR too much. Is this ok with you?

Reply:
Absolutely, I was just offering an alternative to the exception.